### ISSUE TITLE
[New Concept Exercise]: `functools` Module

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#3097]
This issue describes how to implement  the  `functools` module **concept exercise** for the Python track.
The related **concept documents improvement issue** can be found [here](https://github.com/exercism/python/issues/3098).
This issue replaces an earlier issue that has some discussion attached.  The old (closed) issue can be found [here](https://github.com/exercism/python/issues/2366)

<br>

## ✅ Getting started

_If you have not yet created or contributed to a concept exercise, this issue will require some upfront reading to give you the needed background knowledge. Some good example exercises to look at in the repo:_

<details>
<summary>💡<b>Example Exercises</b>💡</summary>
<br>

   1. [Little Sister's Vocabulary](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-vocab)
   2. [Meltdown Mitigation](https://github.com/exercism/python/tree/main/exercises/concept/meltdown-mitigation) 
   3. [Making the Grade](https://github.com/exercism/python/tree/main/exercises/concept/making-the-grade)
   4. [Ellen's Alien Game](https://github.com/exercism/python/tree/main/exercises/concept/ellens-alien-game)

</details>

We also recommend completing one or more of the concept exercises (_they're called "learning exercises"_) [on the website](https://exercism.org/tracks/python/concepts). 

**`Please please read the docs before starting.`** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please go through the following documents:

**General Contributing Docs:**

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | - [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

**Documents on Language Tracks and Concept Exercises**

- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concept Exercises?](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Exercise Stories](https://exercism.org/docs/building/tracks/stories)

## 🎯 Goal

This concept exercise is meant to teach an understanding/use of the `functools` module  in Python. 

<br>

## 💡Learning objectives

Learn more about the functional tools the Python Standard Library provides through the `functools` module.
Build and understanding of and use the following methods and decorators from the module:

- `functools.partial()`
  - `partial.func`
  - `partial.args`
  - `partial.keywords`
-   *class* `functools.partialmethod()` 
-  *class* `functools.singledispatchmethod()`
-  `functools.cache()`
- `@functools.lru_cache()`
- `@functools.singledispatch`
  - `@<function>.register()`
- `@functools.wraps()` -- (_these might move to the `decorators` exercise, depending on how difficult they are to work in_)
   - `functools.update_wrapper()`
-  ~~`@functools.cached_property()`~~ (_this seems better in `class_customization`_)
-  ~~`@functools.total_ordering`~~ (_this method seems more appropriate in the`rich comparisons` exercise_)

<br>

## 🤔 Concepts

- functional tools in python
- `functools` module
- `generic functions`
- `decorators`
- `higher-order functions`
- `partial objects` in python/`partial evaluation` in python
- `single dispatch`

<br>

## 🚫 Topics that are Out of scope

<br>
<details>
<summary><b>Concepts & Subjects that are Out of Scope</b> (<em>click to expand</em>)</summary>
<br>

- `classes`  & `class customization` beyond the direct use of the _class_ methods in this module.
- `comprehensions`
- `comprehensions` in `lambdas`
-  `decorators` (_these have their own exercise.  See issue #2356_ )
- `map()`, `filter()` or `functools.reduce()` in a `comprehension`
-  `functools.reduce()`(_this was already covered with `map()` and `filter()`_)
- `generators`, beyond what is in this module
-  using an `assignment expression` or "walrus" operator (`:=`) in a `lambda`
- class decorators beyond the ones described in this module.
- `enums`

</details>
<br>

## ↩️  Prerequisites

These are the concepts/concept exercises the student should be familiar with before taking on/learning this concept.  
<br>

<details>
<summary>Prereqs  (<em>click to expand</em>)</summary>
<br>

- `basics`
- `bools`
- `classes`
- `class-customization`
- `class-components`
- `comparisons`
- `rich-comparisons`
- `decorators`
- `descriptors`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order-functions`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

</details>

<br>

##  📚 Resources for Writing and Reference

<details>
<summary><b>Resources</b> (<em>click to expand</em>)</summary>
<br>

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs: functional programming HOWTO](https://docs.python.org/3/howto/functional.html)
- [Python Docs: functools module](https://docs.python.org/3/library/functools.html)
- [Pthon Module of the Week:  `functools` - Tools for Maniputlating Fuctions](https://pymotw.com/3/functools/)
- [Florian Dahlitz: Introduction to Python's Functools Module](https://florian-dahlitz.de/blog/introduction-to-functools)
- [PyDanny: Python Partials are Fun!](https://www.pydanny.com/python-partials-are-fun.html)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Trey Hunner:  Ist it a Class or a Function?  It's a callable!](https://treyhunner.com/2019/04/is-it-a-class-or-a-function-its-a-callable/)
- [built-ins: Python Docs](https://docs.python.org/3/library/functions.html)
  - [`map()`](https://docs.python.org/3/library/functions.html#map) & [`filter()`](https://docs.python.org/3/library/functions.html#filter)
  - [`any()`](https://docs.python.org/3/library/functions.html#any) & [`all()`](https://docs.python.org/3/library/functions.html#all)
  - [`min()`](https://docs.python.org/3/library/functions.html#min) & [`max()`](https://docs.python.org/3/library/functions.html#max)
  - [`sum()`](https://docs.python.org/3/library/functions.html#sum) 
- [Real Python: Functional Programming in Python: When and How to Use it](https://realpython.com/python-functional-programming/)

</details>
<br>  

### Exercise Ideas & Stories

Should you need inspiration for an exercise story, you can find a collection [here](https://exercism.org/docs/building/tracks/stories).  You can also port an exercise from another track, **but please make sure to only to include tasks that actually make sense in Python** and that add value for a student.  Remove/replace/add tasks as needed to make the concept clear/workable.

<br>

##  📁 Exercise Files to Be Created

<details open>
<summary><b>File Detail for this Exercise</b></summary>
<table>
<td>
<br>

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This can summarize/paraphrase the linked concept documents if they have already been created (either the `about` or the `introduction`).   The summary does need to have enough information and examples  for the student to complete all the tasks outlined for this concept exercise.

* ### Exercise `instructions.md`

   For more information, see [`instructions.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsinstructionsmd)

   Instructions for an exercise usually center on a story that sets up the code challenge to be solved.  You can create your own story, or fork one from the ones listed [here](https://exercism.org/docs/building/tracks/stories).  Please make sure to give credit to the original authors if you use a story or fork an exercise.

* ### Exercise `Exemplar.py` Solution
  For more information, see [exemplar implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-exemplar-implementation).
  
  This file should not use syntax or datas structures not introduced in this exercise or in this exercise's prerequisites.  It will be used as an "ideal" solution for the challenge, so make sure it conforms to PEP8 and other formatting conventions, and **does not use single letter variable names**.  It should also include proper module and function-level docstrings.  However, it should **NOT** include typehinting or type aliases.

* ### `<Exercise>.py` (Stub) for Implementation
  For more information, see [stub implementation](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-stub-implementation).
  
  This file should provide the expected function names imported for testing, and optionally TODO comments and or docstrings to aid the student in their implementation.  TODOs and docstrings are not required.

* ### `<Exercise>_Test.py` Files
  For more information, see [Tests](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-tests).
  Additionally, please note that Python associates exercise tasks to tests via a [Pytest Marker](https://docs.pytest.org/en/7.1.x/example/markers.html), and uses [`unittest subtests`](https://docs.python.org/3/library/unittest.html#subtests) as a form of test paramaterization.  See the test file for [`Little Sisters Vocab`](https://github.com/exercism/python/blob/main/exercises/concept/little-sisters-vocab/strings_test.py) for examples of how these techniques work.

* ### Exercise `Hints.md`

  For more information on writing hints see [`hints.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  - Hints should provide enough information to get most students "un-stuck" and moving toward a solution.  They should **not** provide a student with a direct solution.
  - You can refer to one or more of the resources linked in this issue above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### Exercise Metadata Files Under  `.meta/config.json`

  For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata-files)

   * `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
   * `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

<br>
</td>
</table>
</details>
<br>

## ♾️ Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [`config.json`](https://exercism.org/docs/building/tracks/config-json).  The track `config.json` file can be found in the root of the Python repo.

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** :  `276fabf5-e7eb-4736-b0fe-a73e02428606`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue    

<br>

## 🎶 Implementation Notes

- As a reminder, code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not strictly require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered either in the introduction to this exercise, or to one of its prerequisites.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste to students, so be as clear as you can_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed.
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest-specific test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

<br>

##  🆘  Next Steps & Getting Help

1.  **`If you'd like to work on this issue, comment saying "I'd like to work on this"`** (_there is no real need to wait for a response, just go ahead, we'll assign you and put a `[claimed]` label on the issue_).
2. If you have any questions while implementing, please post the questions as comments in here, or contact one of the maintainers on our Slack channel.

**Comment by Metallifax** [comment]
Present! Thank you for updating this.

**Comment by BethanyG** [comment]
_naughty bot!_

**Comment by BethanyG** [comment]
@Metallifax -- Just pining here...are you still interested in working on this?
